src/main/scala/firrtl/passes/InferResets.scala

```diff
@@ -0,0 +1,200 @@
+// See LICENSE for license details.
+
+package firrtl.passes
+
+import firrtl._
+import firrtl.ir._
+import firrtl.Mappers._
+import firrtl.traversals.Foreachers._
+import firrtl.analyses.InstanceGraph
+import firrtl.graph.{DiGraph, MutableDiGraph}
+import firrtl.annotations.{ReferenceTarget, TargetToken}
+import firrtl.Utils.toTarget
+import collection.mutable
+
+import scala.util.{Try, Success, Failure}
+
+/** Infers the concrete type of [[Reset]]s by their connections
+  * This is a global inference because ports can be of type [[Reset]]
+  */
+// TODO should we error if a DefMemory is of type AsyncReset? In CheckTypes?
+object InferResets extends Pass {
+
+  final class DifferingDriverTypesException private (msg: String) extends PassException(msg)
+  object DifferingDriverTypesException {
+    def apply(target: ReferenceTarget, tpes: Seq[(Type, Seq[TypeDriver])]): DifferingDriverTypesException = {
+      val xs = tpes.map { case (t, ds) => s"${ds.map(_.target().serialize).mkString(", ")} of type ${t.serialize}" }
+      val msg = s"${target.serialize} driven with multiple types!" + xs.mkString("\n  ", "\n  ", "")
+      new DifferingDriverTypesException(msg)
+    }
+  }
+
+  private sealed trait ResetDriver
```

Comment:
Add high-level overview of what these classes represent.

Reply:
done